### PR TITLE
Add codec and processor combinations to S3 source

### PR DESCRIPTION
**Pull request by Naarcha-AWS**
### Checklist
- [x] By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license and subject to the [Developers Certificate of Origin](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: _data-prepper/common-use-cases/codec-processor-combinations.md [hunk 1]
```diff
@@ -0,0 +1,59 @@
+---
+layout: default
+title: Codec processor combinations
+parent: Common use cases
+nav_order: 25
+---
+
+# Codec processor combinations
+
+At ingestion time, data received by the [`s3 source`]({{site.url}}{{site.baseurl}}/data-prepper/pipelines/configuration/sources/s3/) can be parsed by codecs, such as [Apache Avro](https://avro.apache.org/docs/1.11.1/specification/). which will compress and decompress large sets of data in a certain format before ingestion through a Data Prepper pipelines [processor]({{site.url}}{{site.baseurl}}/data-prepper/pipelines/configuration/processors/processors/).
```
Comment (dlvenable) [inline _data-prepper/common-use-cases/codec-processor-combinations.md:10]: We should only format "s3" as code, not "s3 source." Basically, let's move the back quotes. Thus, this should be:

> by the [`s3` source]

Comment (dlvenable) [inline _data-prepper/common-use-cases/codec-processor-combinations.md:10]: > such as [Apache Avro]

I recommend starting with a codec like JSON over Avro. JSON is probably the most popular codec.

File: _data-prepper/common-use-cases/codec-processor-combinations.md [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+
+## JSON Array
+
+A [JSON array](https://json-schema.org/understanding-json-schema/reference/array) is used to order elements of different types. Because an array is required in JSON, the data contained within the array must be tabular.
+
+The JSON array is most effective with the following codec processor combination. 
+
+| **Compression type** | **Codec** | **Processor** |
+|---|---|---|
+| None /gzip/ Automatic |[JSON]({{site.url}}{{site.baseurl}}/data-prepper/pipelines/configuration/sources/s3#json-codec) | [parse_json]({{site.url}}{{site.baseurl}}data-prepper/pipelines/configuration/processors/parse-json/) |
```
Comment (dlvenable) [inline _data-prepper/common-use-cases/codec-processor-combinations.md:22]: Do we need the compression type column? That is unrelated to the processor/codec.

Comment (dlvenable) [inline _data-prepper/common-use-cases/codec-processor-combinations.md:22]: > [JSON]

Please make this lowercase:

> [json]

Comment (dlvenable) [inline _data-prepper/common-use-cases/codec-processor-combinations.md:22]: When using the `json` codec, the user does not need to use the `parse_json` processor. Indeed, for most use-cases, the user shouldn't use `parse_json`.

File: _data-prepper/common-use-cases/codec-processor-combinations.md [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+
+## NDJSON
+
+Unlike a JSON array, [NDJSON](https://www.npmjs.com/package/ndjson), each row of data can be delimited by a newline, which means data is processed per-line as opposed to an array.
+
+The NDJSON input type is most effective with the following codec processor combination.
+
+| **Compression type** | **Codec** | **Processor** |
+|---|---|---|---|
+| None /gzip/ Automatic | [Newline]({{site.url}}{{site.baseurl}}/data-prepper/pipelines/configuration/sources/s3#newline-codec)  |
```
Comment (dlvenable) [inline _data-prepper/common-use-cases/codec-processor-combinations.md:32]: > [Newline]

This should be all lowercase:

> [newline]

Comment (dlvenable) [inline _data-prepper/common-use-cases/codec-processor-combinations.md:32]: When using the `newline` to read ND-JSON data, users need to use the `parse_json` processor. So that should be added here.

File: _data-prepper/common-use-cases/codec-processor-combinations.md [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+
+## CSV
+
+The CSV data type inputs data as a table. It can used without a codec or processor, but it does require the one or the other. 
+
+The CSV input type is most effective with the following codec processor combinations.
+
+| **Compression type** | **Codec** | **Processor** | **Limitations**
+|---|---|---|---|---|
+| None /gzip/ Automatic | [CSV]({{site.url}}{{site.baseurl}}/data-prepper/pipelines/configuration/sources/s3#csv-codec) | None | Header is automatically detected and used for index mapping with the CSV codec. |
```
Comment (dlvenable) [inline _data-prepper/common-use-cases/codec-processor-combinations.md:42]: > [CSV]

This should be lowercase:

> [csv]

File: _data-prepper/common-use-cases/codec-processor-combinations.md [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+| None /gzip/ Automatic | [Newline]({{site.url}}{{site.baseurl}}/data-prepper/pipelines/configuration/sources/s3#newline-codec) | [CSV]({{site.url}}{{site.baseurl}}/data-prepper/pipelines/configuration/processors/csv/) | Header is detected if `header_destination` is configured in the `newline` codec. <br>The header can be used in the CSV processor under `column_names_source_key`. |
```
Comment (dlvenable) [inline _data-prepper/common-use-cases/codec-processor-combinations.md:43]: Again, let's be consistent with the case:

> [Newline]

to:

> [newline]

And:

> [CSV]

to:

> [csv]

File: _data-prepper/common-use-cases/codec-processor-combinations.md [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+|---|---|---|---|
+| None /gzip/ Automatic | [Newline]({{site.url}}{{site.baseurl}}/data-prepper/pipelines/configuration/sources/s3#newline-codec)  | 
+
+## CSV
+
+The CSV data type inputs data as a table. It can used without a codec or processor, but it does require the one or the other. 
+
+The CSV input type is most effective with the following codec processor combinations.
+
+| **Compression type** | **Codec** | **Processor** | **Limitations**
```
Comment (dlvenable) [inline _data-prepper/common-use-cases/codec-processor-combinations.md:40]: I wonder if the table format is the right way to convey this to users. It seems that we could have the following instead.

1. Some example data
2. A paragraph on the data and the configuration
3. A sample pipeline configuration, or perhaps just a partial configuration.

File: _data-prepper/common-use-cases/codec-processor-combinations.md [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+| None /gzip/ Automatic | None | [CSV]({{site.url}}{{site.baseurl}}/data-prepper/pipelines/configuration/processors/csv/) | N/A |
+
+
+## Parquet
+
+[Apache Parquet](https://parquet.apache.org/docs/overview/) is columnar storage format built for Hadoop. It is most efficient without the use of a codec, however positive results can be achieved when paired with the following processors depending on the compression type.
+
+| **Compression type** | **Processor** |
+|---|---|
+| None /gzip/ Automatic | [Parquet]({{site.url}}{{site.baseurl}}/data-prepper/pipelines/configuration/sinks/s3#parquet-codec) |
```
Comment (dlvenable) [inline _data-prepper/common-use-cases/codec-processor-combinations.md:53]: I think that a table may not convey the best way to convey this. For one, using the `parquet` codec versus using `s3_select` is a decision that may require some additional consideration. And the way you configure `s3_select` is different.

File: _data-prepper/pipelines/configuration/sources/s3.md [hunk 8]
```diff
@@ -149,6 +149,22 @@ Option | Required | Type   | Description
 `header` | No   | String list | The header containing the column names used to parse CSV data.
 `detect_header` | No   | Boolean   | Whether the first line of the S3 object should be interpreted as a header. Default is `true`.
 
+## Codec and processor combinations 
+
+At ingestion time, data received by the S3 source can be parsed by codecs. However, codecs only support a particular type of data input type, compressions types, and processors. The following table describes each codec, data type, and processor combination where performance of the pipeline can be improved. 
+
+| **Input Data Type** | **Compression type** | **Codec** | **Processor** | Notes |
```
Comment (dlvenable) [inline _data-prepper/pipelines/configuration/sources/s3.md:156]: Rather than adding a table here, perhaps we can link to the other page for users.

**Review by dlvenable — CHANGES_REQUESTED** [review]

File: _data-prepper/common-use-cases/codec-processor-combinations.md (hunk 1 above)
Comment (Naarcha-AWS) [inline _data-prepper/common-use-cases/codec-processor-combinations.md:10]:
```suggestion
At ingestion time, data received by the [s3 source]({{site.url}}{{site.baseurl}}/data-prepper/pipelines/configuration/sources/s3/) can be parsed by codecs, such as [Apache Avro](https://avro.apache.org/docs/1.11.1/specification/). which will compress and decompress large sets of data in a certain format before ingestion through a Data Prepper pipelines [processor]({{site.url}}{{site.baseurl}}/data-prepper/pipelines/configuration/processors/processors/).
```

File: _data-prepper/common-use-cases/codec-processor-combinations.md (hunk 2 above)
Comment (Naarcha-AWS) [inline _data-prepper/common-use-cases/codec-processor-combinations.md:22]: Do we have JSON processor besides `parse_JSON`?

Comment (dlvenable) [reply _data-prepper/common-use-cases/codec-processor-combinations.md:22]: No, `parse_json` is the only JSON processor we have.

Comment (Naarcha-AWS) [reply _data-prepper/common-use-cases/codec-processor-combinations.md:22]: So do they need a processor at all when using the JSON codec?

File: _data-prepper/pipelines/configuration/sources/s3.md (hunk 8 above)
Comment (Naarcha-AWS) [reply _data-prepper/pipelines/configuration/sources/s3.md:156]: I'm going to remove the table.

File: _data-prepper/common-use-cases/codec-processor-combinations.md (hunk 2 above)
Comment (dlvenable) [reply _data-prepper/common-use-cases/codec-processor-combinations.md:22]: When using the `json` codec, users don't need a processor. Their data will already be somewhat ready.

Keep in mind that customers may wish to use any number of processors after the data has been parsed to normalize data, remove data, etc.

File: _data-prepper/common-use-cases/codec-processor-combinations.md (hunk 7 above)
Comment (Naarcha-AWS) [reply _data-prepper/common-use-cases/codec-processor-combinations.md:53]: What additional considerations?

**Comment by Naarcha-AWS** [comment]
@dlvenable: Addressed most of your feedback. I'll need some help developing examples, but if we want to marge this now before we add them we can circle back around to the examples.

File: _data-prepper/common-use-cases/codec-processor-combinations.md [hunk 9]
```diff
@@ -0,0 +1,55 @@
...
+
+## Parquet
+
+[Apache Parquet](https://parquet.apache.org/docs/overview/) is columnar storage format built for Hadoop. It is most efficient without the use of a codec, however positive results can be achieved when paired with the following processors depending on the compression type.
+
+
+- [Parquet]({{site.url}}{{site.baseurl}}/data-prepper/pipelines/configuration/sinks/s3#parquet-codec) |
+- [S3 Select]({{site.url}}{{site.baseurl}}/data-prepper/pipelines/configuration/sources/s3#using-s3_select-with-the-s3-source) |
+
+## Avro
```
Comment (Naarcha-AWS) [inline _data-prepper/common-use-cases/codec-processor-combinations.md:52]:
```suggestion
## Avro
```

File: _data-prepper/common-use-cases/codec-processor-combinations.md [hunk 10]
```diff
@@ -0,0 +1,55 @@
...
+
+[Apace Avro] helps streamline streaming data pipelines. It is most efficient when used with the [Avro]({{site.url}}{{site.baseurl}}/data-prepper/pipelines/configuration/sinks/s3#avro-codec) inside an `s3` sink.
```
Comment (Naarcha-AWS) [inline _data-prepper/common-use-cases/codec-processor-combinations.md:54]:
```suggestion
[Apache Avro] helps streamline streaming data pipelines. It is most efficient when used with the [`avro`]({{site.url}}{{site.baseurl}}/data-prepper/pipelines/configuration/sinks/s3#avro-codec) codec inside an `s3` sink.
```

File: _data-prepper/common-use-cases/codec-processor-combinations.md [hunk 11]
```diff
@@ -0,0 +1,55 @@
...
+### csv codec
+
+The [csv]({{site.url}}{{site.baseurl}}/data-prepper/pipelines/configuration/sources/s3#csv-codec) codec without a processor. When used without a processor, the `csv` codec automatically detects headers from the CSV and uses them for index mapping.
+
+### newline codec 
+
+The [newline]({{site.url}}{{site.baseurl}}/data-prepper/pipelines/configuration/sources/s3#newline-codec) codec parses each row as a single log event. However, the codec will only detect a header when `header_destination` is configured. Then, the [csv]({{site.url}}{{site.baseurl}}/data-prepper/pipelines/configuration/processors/csv/) processor outputs the event into columns. The header detected in `header_destination` from the newline codec can be used in the `csv` processor under `column_names_source_key.`
+
+
+## Parquet
```
Comment (Naarcha-AWS) [inline _data-prepper/common-use-cases/codec-processor-combinations.md:44]:
```suggestion
## parquet
```

File: _data-prepper/common-use-cases/codec-processor-combinations.md [hunk 12]
```diff
@@ -0,0 +1,55 @@
...
+
+[Apache Parquet](https://parquet.apache.org/docs/overview/) is columnar storage format built for Hadoop. It is most efficient without the use of a codec, however positive results can be achieved when paired with the following processors depending on the compression type.
```
Comment (Naarcha-AWS) [inline _data-prepper/common-use-cases/codec-processor-combinations.md:46]:
```suggestion
[Apache Parquet](https://parquet.apache.org/docs/overview/) is a columnar storage format built for Hadoop. It is most efficient without the use of a codec, however positive results can be achieved when paired with the following processors depending on the compression type.
```

File: _data-prepper/common-use-cases/codec-processor-combinations.md [hunk 13]
```diff
@@ -0,0 +1,55 @@
...
+
+
+- [Parquet]({{site.url}}{{site.baseurl}}/data-prepper/pipelines/configuration/sinks/s3#parquet-codec) |
```
Comment (Naarcha-AWS) [inline _data-prepper/common-use-cases/codec-processor-combinations.md:49]:
```suggestion

```

File: _data-prepper/common-use-cases/codec-processor-combinations.md [hunk 14]
```diff
@@ -0,0 +1,55 @@
...
+- [S3 Select]({{site.url}}{{site.baseurl}}/data-prepper/pipelines/configuration/sources/s3#using-s3_select-with-the-s3-source) |
```
Comment (Naarcha-AWS) [inline _data-prepper/common-use-cases/codec-processor-combinations.md:50]:
```suggestion

```

File: _data-prepper/common-use-cases/codec-processor-combinations.md [hunk 15]
```diff
@@ -0,0 +1,55 @@
+---
+layout: default
+title: Codec processor combinations
+parent: Common use cases
+nav_order: 25
+---
+
+# Codec processor combinations
+
+At ingestion time, data received by the [`s3` source]({{site.url}}{{site.baseurl}}/data-prepper/pipelines/configuration/sources/s3/) can be parsed by [codecs]({{site.url}}{{site.baseurl}}/data-prepper/pipelines/configuration/sources/s3#codec). which will compress and decompress large sets of data in a certain format before ingestion through a Data Prepper pipelines [processor]({{site.url}}{{site.baseurl}}/data-prepper/pipelines/configuration/processors/processors/).
```
Comment (vagimeli) [inline _data-prepper/common-use-cases/codec-processor-combinations.md:10]:
```suggestion
At ingestion time, data received by the [`s3` source]({{site.url}}{{site.baseurl}}/data-prepper/pipelines/configuration/sources/s3/) can be parsed by [codecs]({{site.url}}{{site.baseurl}}/data-prepper/pipelines/configuration/sources/s3#codec). Codecs compresses and decompresses large data sets in a certain format before ingestion them through a Data Prepper pipeline [processor]({{site.url}}{{site.baseurl}}/data-prepper/pipelines/configuration/processors/processors/).
```

File: _data-prepper/common-use-cases/codec-processor-combinations.md [hunk 16]
```diff
@@ -0,0 +1,55 @@
...
+
+While most codecs can be used with most processors, the following codec processor combinations can help make your pipeline more efficient when using the following input types.
```
Comment (vagimeli) [inline _data-prepper/common-use-cases/codec-processor-combinations.md:12]:
```suggestion
While most codecs can be used with most processors, the following codec processor combinations can make your pipeline more efficient when used with the following input types.
```

File: _data-prepper/common-use-cases/codec-processor-combinations.md [hunk 17]
```diff
@@ -0,0 +1,55 @@
...
+
+## JSON Array
```
Comment (vagimeli) [inline _data-prepper/common-use-cases/codec-processor-combinations.md:14]:
```suggestion
## JSON array
```

File: _data-prepper/common-use-cases/codec-processor-combinations.md [hunk 18]
```diff
@@ -0,0 +1,55 @@
...
+
+A [JSON array](https://json-schema.org/understanding-json-schema/reference/array) is used to order elements of different types. Because an array is required in JSON, the data contained within the array must be tabular.
+
+The JSON array does not require a processor. 
+
+## NDJSON
+
+Unlike a JSON array, [NDJSON](https://www.npmjs.com/package/ndjson), each row of data can be delimited by a newline, which means data is processed per-line as opposed to an array.
```
Comment (vagimeli) [inline _data-prepper/common-use-cases/codec-processor-combinations.md:22]:
```suggestion
Unlike a JSON array, [NDJSON](https://www.npmjs.com/package/ndjson) allows for each row of data to be delimited by a newline, meaning data is processed per line instead of an array.
```

File: _data-prepper/common-use-cases/codec-processor-combinations.md [hunk 19]
```diff
@@ -0,0 +1,53 @@
...
+
+The NDJSON input type requires is parsed using the [newline]({{site.url}}{{site.baseurl}}/data-prepper/pipelines/configuration/sources/s3#newline-codec) codec, which parses each single line as a single log event. The [parse_json]({{site.url}}{{site.baseurl}}data-prepper/pipelines/configuration/processors/parse-json/) processor then outputs each line as a single event.
```
Comment (vagimeli) [inline _data-prepper/common-use-cases/codec-processor-combinations.md:24]:
```suggestion
The NDJSON input type is parsed using the [newline]({{site.url}}{{site.baseurl}}/data-prepper/pipelines/configuration/sources/s3#newline-codec) codec, which parses each single line as a single log event. The [parse_json]({{site.url}}{{site.baseurl}}data-prepper/pipelines/configuration/processors/parse-json/) processor then outputs each line as a single event.
```

File: _data-prepper/common-use-cases/codec-processor-combinations.md [hunk 20]
```diff
@@ -0,0 +1,53 @@
...
+
```
Comment (vagimeli) [inline _data-prepper/common-use-cases/codec-processor-combinations.md:25]: Delete extra blank lines

File: _data-prepper/common-use-cases/codec-processor-combinations.md [hunk 21]
```diff
@@ -0,0 +1,53 @@
...
+
+
+## CSV
+
+The CSV data type inputs data as a table. It can used without a codec or processor, but it does require one or the other, for example, either just the `csv` processors or the `csv` codec.
```
Comment (vagimeli) [inline _data-prepper/common-use-cases/codec-processor-combinations.md:30]:
```suggestion
The CSV data type inputs data as a table. It can used without a codec or processor, but it does require one or the other, for example, either just the `csv` processor or the `csv` codec.
```

File: _data-prepper/common-use-cases/codec-processor-combinations.md [hunk 22]
```diff
@@ -0,0 +1,53 @@
...
+
+The CSV input type is most effective with the following codec/processor combinations.
```
Comment (vagimeli) [inline _data-prepper/common-use-cases/codec-processor-combinations.md:32]:
```suggestion
The CSV input type is most effective when used with the following codec processor combinations.
```

File: _data-prepper/common-use-cases/codec-processor-combinations.md [hunk 23]
```diff
@@ -0,0 +1,53 @@
...
+
+
+### csv codec
```
Comment (vagimeli) [inline _data-prepper/common-use-cases/codec-processor-combinations.md:35]:
```suggestion
### CSV codec
```

File: _data-prepper/common-use-cases/codec-processor-combinations.md [hunk 24]
```diff
@@ -0,0 +1,53 @@
...
+
+The [csv]({{site.url}}{{site.baseurl}}/data-prepper/pipelines/configuration/sources/s3#csv-codec) codec without a processor. When used without a processor, the `csv` codec automatically detects headers from the CSV and uses them for index mapping.
```
Comment (vagimeli) [inline _data-prepper/common-use-cases/codec-processor-combinations.md:37]:
```suggestion
When the [`csv` codec]({{site.url}}{{site.baseurl}}/data-prepper/pipelines/configuration/sources/s3#csv-codec) is used without a processor, it automatically detects headers from the CSV and uses them for index mapping.
```

Comment (vagimeli) [reply _data-prepper/common-use-cases/codec-processor-combinations.md:37]: Should this read "...the CSV **file** and..."

File: _data-prepper/common-use-cases/codec-processor-combinations.md [hunk 25]
```diff
@@ -0,0 +1,53 @@
...
+
+### newline codec
```
Comment (vagimeli) [inline _data-prepper/common-use-cases/codec-processor-combinations.md:39]:
```suggestion
### Newline codec 
```

File: _data-prepper/common-use-cases/codec-processor-combinations.md [hunk 26]
```diff
@@ -0,0 +1,53 @@
...
+The CSV input type is most effective with the following codec/processor combinations.
+
+
+### csv codec
+
+The [csv]({{site.url}}{{site.baseurl}}/data-prepper/pipelines/configuration/sources/s3#csv-codec) codec without a processor. When used without a processor, the `csv` codec automatically detects headers from the CSV and uses them for index mapping.
+
+### newline codec 
+
+The [newline]({{site.url}}{{site.baseurl}}/data-prepper/pipelines/configuration/sources/s3#newline-codec) codec parses each row as a single log event. However, the codec will only detect a header when `header_destination` is configured. Then, the [csv]({{site.url}}{{site.baseurl}}/data-prepper/pipelines/configuration/processors/csv/) processor outputs the event into columns. The header detected in `header_destination` from the newline codec can be used in the `csv` processor under `column_names_source_key.`
```
Comment (vagimeli) [inline _data-prepper/common-use-cases/codec-processor-combinations.md:41]:
```suggestion
The [`newline` codec]({{site.url}}{{site.baseurl}}/data-prepper/pipelines/configuration/sources/s3#newline-codec) parses each row as a single log event. The codec will only detect a header when `header_destination` is configured. The processor then [csv]({{site.url}}{{site.baseurl}}/data-prepper/pipelines/configuration/processors/csv/) outputs the event into columns. The header detected in `header_destination` from the `newline` codec can be used in the `csv` processor under `column_names_source_key.`
```

File: _data-prepper/common-use-cases/codec-processor-combinations.md (hunk 11 above)
Comment (vagimeli) [reply _data-prepper/common-use-cases/codec-processor-combinations.md:44]: Leave the headings capitalized. Use code font in the body text.

File: _data-prepper/common-use-cases/codec-processor-combinations.md [hunk 27]
```diff
@@ -0,0 +1,53 @@
...
+
+
+## Parquet
+
+[Apache Parquet](https://parquet.apache.org/docs/overview/) is a columnar storage format built for Hadoop. It is most efficient without the use of a codec, however positive results can be achieved when configured with [S3 Select]({{site.url}}{{site.baseurl}}/data-prepper/pipelines/configuration/sources/s3#using-s3_select-with-the-s3-source).
```
Comment (vagimeli) [inline _data-prepper/common-use-cases/codec-processor-combinations.md:46]:
```suggestion
[Apache Parquet](https://parquet.apache.org/docs/overview/) is a columnar storage format built for Hadoop. It is most efficient without the use of a codec. Positive results, however, can be achieved when it's configured with [S3 Select]({{site.url}}{{site.baseurl}}/data-prepper/pipelines/configuration/sources/s3#using-s3_select-with-the-s3-source).
```

File: _data-prepper/common-use-cases/codec-processor-combinations.md [hunk 28]
```diff
@@ -0,0 +1,53 @@
...
+
+
+
+## Avro
+
+[Apache Avro] helps streamline streaming data pipelines. It is most efficient when used with the [`avro`]({{site.url}}{{site.baseurl}}/data-prepper/pipelines/configuration/sinks/s3#avro-codec) codec inside an `s3` sink.
```
Comment (vagimeli) [inline _data-prepper/common-use-cases/codec-processor-combinations.md:52]:
```suggestion
[Apache Avro] helps streamline streaming data pipelines. It is most efficient when used with the [`avro` codec]({{site.url}}{{site.baseurl}}/data-prepper/pipelines/configuration/sinks/s3#avro-codec) inside an `s3` sink.
```

File: _data-prepper/pipelines/configuration/sources/s3.md [hunk 29]
```diff
@@ -121,7 +121,7 @@ Option | Required | Type | Description
 
 ## codec
```
Comment (vagimeli) [inline _data-prepper/pipelines/configuration/sources/s3.md:122]:
```suggestion
## Codec
```

File: _data-prepper/pipelines/configuration/sources/s3.md [hunk 30]
```diff
@@ -121,7 +121,7 @@ Option | Required | Type | Description
...
 
-The `codec` determines how the `s3` source parses each S3 object.
+The `codec` determines how the `s3` source parses each S3 object. For increased and more efficient performance, you can use codecs in [combination]({{site.url}}{{site.baseurl}}/data-prepper/common-use-cases/codec-processor-combinations/) with certain processors.
```
Comment (vagimeli) [inline _data-prepper/pipelines/configuration/sources/s3.md:124]:
```suggestion
The `codec` determines how the `s3` source parses each Amazon S3 object. For increased and more efficient performance, you can use [codec combinations]({{site.url}}{{site.baseurl}}/data-prepper/common-use-cases/codec-processor-combinations/) with certain processors.
```

File: _data-prepper/pipelines/configuration/sources/s3.md [hunk 31]
```diff
@@ -121,7 +121,7 @@ Option | Required | Type | Description
...
 
 ### newline codec
```
Comment (vagimeli) [inline _data-prepper/pipelines/configuration/sources/s3.md:126]:
```suggestion
### Newline codec
```

File: _data-prepper/pipelines/configuration/sources/s3.md [hunk 32]
```diff
@@ -149,6 +149,9 @@ Option | Required | Type   | Description
 `header` | No   | String list | The header containing the column names used to parse CSV data.
 `detect_header` | No   | Boolean   | Whether the first line of the S3 object should be interpreted as a header. Default is `true`.
```
Comment (vagimeli) [inline _data-prepper/pipelines/configuration/sources/s3.md:150]:
```suggestion
`detect_header` | No   | Boolean   | Whether the first line of the Amazon S3 object should be interpreted as a header. Default is `true`.
```

File: _data-prepper/pipelines/configuration/sources/s3.md [hunk 33]
```diff
@@ -149,6 +149,9 @@ Option | Required | Type   | Description
...
 
+
+
+
 ## Using `s3_select` with the `s3` source<a name="s3_select"></a>
 
 When configuring `s3_select` to parse S3 objects, use the following options.
```
Comment (vagimeli) [inline _data-prepper/pipelines/configuration/sources/s3.md:157]:
```suggestion
When configuring `s3_select` to parse Amazon S3 objects, use the following options:
```

File: _data-prepper/pipelines/configuration/sources/s3.md [hunk 34]
```diff
@@ -225,6 +228,7 @@ Option | Required | Type | Description
 `interval` | Yes  | String | Indicates the minimum interval between each scan. The next scan in the interval will start after the interval duration from the last scan ends and when all the objects from the previous scan are processed. Supports ISO_8601 notation strings, such as `PT20.345S` or `PT15M`, and notation strings for seconds (`60s`) and milliseconds (`1600ms`).
```
Comment (vagimeli) [inline _data-prepper/pipelines/configuration/sources/s3.md:228]:
```suggestion
`interval` | Yes  | String | Indicates the minimum interval between each scan. The next scan in the interval will start after the interval duration from the last scan ends and when all the objects from the previous scan are processed. Supports ISO 8601 notation strings, such as `PT20.345S` or `PT15M`, and notation strings for seconds (`60s`) and milliseconds (`1600ms`). 
```

File: _data-prepper/pipelines/configuration/sources/s3.md [hunk 35]
```diff
@@ -225,6 +228,7 @@ Option | Required | Type | Description
 `interval` | Yes  | String | Indicates the minimum interval between each scan. The next scan in the interval will start after the interval duration from the last scan ends and when all the objects from the previous scan are processed. Supports ISO_8601 notation strings, such as `PT20.345S` or `PT15M`, and notation strings for seconds (`60s`) and milliseconds (`1600ms`). 
 `count` | No | Integer | Specifies how many times a bucket will be scanned. Defaults to `Integer.MAX_VALUE`.
```
Comment (vagimeli) [inline _data-prepper/pipelines/configuration/sources/s3.md:229]:
```suggestion
`count` | No | Integer | Specifies how many times a bucket is scanned. Defaults to `Integer.MAX_VALUE`.
```

File: _data-prepper/pipelines/configuration/sources/s3.md [hunk 36]
```diff
@@ -225,6 +228,7 @@ Option | Required | Type | Description
...
 
+
 ## Metrics
 
 The `s3` source includes the following metrics.
```
Comment (vagimeli) [inline _data-prepper/pipelines/configuration/sources/s3.md:234]:
```suggestion
The `s3` source includes the following metrics:
```

**Review by vagimeli — COMMENTED** [review]
@Naarcha-AWS Editorial review is complete. Please see edits and ping me directly once they've been addressed so that I can approve for merging. Thanks!

**Review by dlvenable — APPROVED** [review]
I think we should include some examples in the `codec-processor-combinations` file, but we can follow on with this.

File: _data-prepper/common-use-cases/codec-processor-combinations.md (hunk 23 above)
Comment (Naarcha-AWS) [reply _data-prepper/common-use-cases/codec-processor-combinations.md:35]: This is meant to match the codec appears in the code. Its the standard the Data Prepper team wishes to use and is used across the entire Data Prepper guide.

File: _data-prepper/pipelines/configuration/sources/s3.md (hunk 31 above)
Comment (Naarcha-AWS) [reply _data-prepper/pipelines/configuration/sources/s3.md:126]: See previous comments about headers in the Data Prepper section.

File: _data-prepper/pipelines/configuration/sources/s3.md (hunk 35 above)
Comment (Naarcha-AWS) [reply _data-prepper/pipelines/configuration/sources/s3.md:229]: The `count` doesn't occur until the pipeline runs, hence the tense.

File: _data-prepper/pipelines/configuration/sources/s3.md (hunk 36 above)
Comment (Naarcha-AWS) [reply _data-prepper/pipelines/configuration/sources/s3.md:234]: I was told by @natebower to make this a period.

File: _data-prepper/pipelines/configuration/sources/s3.md (hunk 29 above)
Comment (Naarcha-AWS) [reply _data-prepper/pipelines/configuration/sources/s3.md:122]: See previous comment about headers.

File: _data-prepper/common-use-cases/codec-processor-combinations.md [hunk 37]
```diff
@@ -0,0 +1,53 @@
...
+The CSV data type inputs data as a table. It can used without a codec or processor, but it does require one or the other, for example, either just the `csv` processor or the `csv` codec.
+
+The CSV input type is most effective when used with the following codec processor combinations.
+
+
+### csv codec
+
+When the [`csv` codec]({{site.url}}{{site.baseurl}}/data-prepper/pipelines/configuration/sources/s3#csv-codec) is used without a processor, it automatically detects headers from the CSV and uses them for index mapping.
+
+### Newline codec
```
Comment (Naarcha-AWS) [inline _data-prepper/common-use-cases/codec-processor-combinations.md:39]:
```suggestion
### `newline` codec 
```

File: _data-prepper/common-use-cases/codec-processor-combinations.md (hunk 11 above)
Comment (Naarcha-AWS) [reply _data-prepper/common-use-cases/codec-processor-combinations.md:44]: Parquet and Avro are the exceptions to the code rule since they represent a data storage format, not something that appears in code.

File: _data-prepper/common-use-cases/codec-processor-combinations.md (hunk 20 above)
Comment (vagimeli) [reply _data-prepper/common-use-cases/codec-processor-combinations.md:25]: @Naarcha-AWS The comments don't appear to be addressed. This file has several sections with extra blank lines that need to be deleted so that this document is in proper format.

File: _data-prepper/common-use-cases/codec-processor-combinations.md [hunk 38]
```diff
@@ -0,0 +1,53 @@
...
+
+### `newline` codec 
+
+The [`newline` codec]({{site.url}}{{site.baseurl}}/data-prepper/pipelines/configuration/sources/s3#newline-codec) parses each row as a single log event. The codec will only detect a header when `header_destination` is configured. The processor then [csv]({{site.url}}{{site.baseurl}}/data-prepper/pipelines/configuration/processors/csv/) outputs the event into columns. The header detected in `header_destination` from the `newline` codec can be used in the `csv` processor under `column_names_source_key.`
+
+
+## Parquet
+
+[Apache Parquet](https://parquet.apache.org/docs/overview/) is a columnar storage format built for Hadoop. It is most efficient without the use of a codec. Positive results, however, can be achieved when it's configured with [S3 Select]({{site.url}}{{site.baseurl}}/data-prepper/pipelines/configuration/sources/s3#using-s3_select-with-the-s3-source).
+
```
Comment (vagimeli) [inline _data-prepper/common-use-cases/codec-processor-combinations.md:47]: @Naarcha-AWS Delete extra blank lines

**Review by vagimeli — APPROVED** [review]
@Naarcha-AWS Check the file to delete the extra blank lines so that it's properly formatted before merging. Thanks!

File: _data-prepper/common-use-cases/codec-processor-combinations.md (hunk 6 above)
Comment (vagimeli) [reply _data-prepper/common-use-cases/codec-processor-combinations.md]: @Naarcha-AWS Did this comment from @dlvenable get addressed? He needs to write this section based on his comments, as they are technical in nature and he is the SME.

**Comment by vagimeli** [comment]
@Naarcha-AWS If you all add more text in the examples, those changes need to go through editorial review.
